frontend/src/pages/[username]/[projectName]/index.js

```diff
@@ -0,0 +1,413 @@
+// TODO: this page became monolithic, it needs global refactoring.
+import React, { useEffect, useState, useContext } from 'react'
+import { useRouter } from 'next/router'
+import { uniqBy } from 'lodash'
+import {
+  Button,
+  Form,
+  Input,
+  Loader,
+  Message,
+  Segment,
+  TextArea,
+} from 'semantic-ui-react'
+
+import { Page } from '@components/Page'
+import useForm from '@hooks/useForm'
+import { ProjectUpdateFormModel } from '@models/ProjectUpdateForm'
+import { pollMigrationStatus, useDefaultBranchFiles, useRepo } from '@hooks/Gitea'
+import { commitFiles } from '@utils/giteaInternalApi'
+import {
+  getDefaultBranchFiles,
+  getRepo,
+  repoExists,
+  updateRepo,
+} from '@utils/giteaApi'
+import { findReadme, renderReadme } from '@utils/index'
+import {
+  getBoardInfo,
+  getBoardZipInfo,
+  getKitspaceYAMLJson,
+  hasInteractiveBom,
+} from '@utils/projectPage'
+import { AuthContext } from '@contexts/AuthContext'
+import ErrorPage from '@pages/_error'
+import BoardShowcase from '@components/Board/BoardShowcase'
+import BoardExtraMenus from '@components/Board/BoardExtrasMenu'
+import OrderPCBs from '@components/Board/OrderPCBs'
+import BuyParts from '@components/Board/BuyParts/index'
+import InfoBar from '@components/Board/InfoBar'
+import Readme from '@components/Board/Readme'
+import UploadModal from '@components/UploadModal'
+
+export const getServerSideProps = async ({ params, query, req }) => {
+  const processorUrl = process.env.KITSPACE_PROCESSOR_URL
+  const repoFullname = `${params.username}/${params.projectName}`
+  const assetsPath = `${processorUrl}/files/${repoFullname}/HEAD`
+
+  // Only the repo owner can upload files.
+  const hasUploadPermission = params.username === req?.session?.user?.username
```

Comment:
I think we should be checking the Gitea API for if the user has write permission to this project.

Reply:
Resolved via e998119.